### PR TITLE
[Refactor] refactor knn

### PR DESCRIPTION
**Pull request by Jiahao000**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by fangyixiao18** [comment]
LGTM

File: tools/benchmarks/classification/knn_imagenet/test_knn.py [hunk 1]
```diff
@@ -145,30 +139,41 @@ def main():
...
+        seed=args.seed,
+        dist_mode=distributed,
+        pool_cfg=copy.deepcopy(dataset_cfg.pool_cfg))
+    extractor_val = Extractor(
+        extract_dataloader=data_loader_val,
+        seed=args.seed,
+        dist_mode=distributed,
+        pool_cfg=copy.deepcopy(dataset_cfg.pool_cfg))
+    train_feats = extractor_train(model)['feat5']
+    val_feats = extractor_val(model)['feat5']
```
Comment (fangyixiao18) [inline tools/benchmarks/classification/knn_imagenet/test_knn.py:166]: Is it better not to write hard code here?

File: tools/benchmarks/classification/knn_imagenet/slurm_test_knn_pretrain.sh [hunk 2]
```diff
@@ -11,7 +11,6 @@ PY_ARGS=${@:5}
 GPUS=${GPUS:-8}
 GPUS_PER_NODE=${GPUS_PER_NODE:-8}
 CPUS_PER_TASK=${CPUS_PER_TASK:-5}
-PORT=${PORT:-29500}
 SRUN_ARGS=${SRUN_ARGS:-""}
 
 # set work_dir according to config path and pretrained model to distinguish different models
```
Comment (fangyixiao18) [inline tools/benchmarks/classification/knn_imagenet/slurm_test_knn_pretrain.sh:16]: better to set WORK_DIR by users

File: tools/benchmarks/classification/knn_imagenet/slurm_test_knn_epoch.sh [hunk 3]
```diff
@@ -32,5 +31,4 @@ srun -p ${PARTITION} \
     ${SRUN_ARGS} \
     python -u tools/benchmarks/classification/knn_imagenet/test_knn.py $CFG \
         --checkpoint $WORK_DIR/epoch_${EPOCH}.pth \
```
Comment (fangyixiao18) [inline tools/benchmarks/classification/knn_imagenet/slurm_test_knn_epoch.sh:33]: better to set checkpoint path directly according to command line, not by epoch num

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: tools/benchmarks/classification/knn_imagenet/slurm_test_knn_pretrain.sh (hunk 2 above)
Comment (Jiahao000) [reply tools/benchmarks/classification/knn_imagenet/slurm_test_knn_pretrain.sh:16]: The script style follows and is consistent with the `svm_voc07` benchmark.

File: tools/benchmarks/classification/knn_imagenet/slurm_test_knn_epoch.sh (hunk 3 above)
Comment (Jiahao000) [reply tools/benchmarks/classification/knn_imagenet/slurm_test_knn_epoch.sh:33]: The script style follows and is consistent with the `svm_voc07` benchmark.

File: tools/benchmarks/classification/knn_imagenet/test_knn.py (hunk 1 above)
Comment (Jiahao000) [reply tools/benchmarks/classification/knn_imagenet/test_knn.py:166]: We only need the last-stage avgpooled feature here.

**Review by fangyixiao18 — APPROVED** [review]

**Pull request merged**